src/hotspot/share/opto/mulnode.cpp

```diff
@@ -1683,3 +1713,64 @@ const Type* RotateRightNode::Value(PhaseGVN* phase) const {
     return TypeLong::LONG;
   }
 }
+
+// Helper method to transform:
+// patterns similar to (v << 2) & 3 to 0
+// and
+// patterns similar to (v1 + (v2 << 2)) & 3 transformed to v1 & 3
+bool MulNode::AndIL_shift_and_mask(PhaseGVN* phase, Node* mask, Node* shift, BasicType bt) {
+  if (mask == NULL || shift == NULL) {
+    return false;
+  }
+  const TypeInteger* mask_t = phase->type(mask)->isa_integer(bt);
+  const TypeInteger* shift_t = phase->type(shift)->isa_integer(bt);
+  if (mask_t == NULL || shift_t == NULL) {
+    return false;
+  }
+  if (bt == T_LONG && shift != NULL && shift->Opcode() == Op_ConvI2L) {
+    bt = T_INT;
+    shift = shift->in(1);
+    if (shift == NULL) {
+      return false;
+    }
+  }
+  if (shift->Opcode() != Op_LShift(bt)) {
+    return false;
+  }
+  Node* shift2 = shift->in(2);
+  if (shift2 == NULL) {
+    return false;
+  }
+  const Type* shift2_t = phase->type(shift2);
+  if (!shift2_t->isa_int() || !shift2_t->is_int()->is_con()) {
+    return false;
+  }
+
+  jint shift_con = shift2_t->is_int()->get_con() & ((bt == T_INT ? BitsPerJavaInteger : BitsPerJavaLong) - 1);
+  if ((((jlong)1) << shift_con) > mask_t->hi_as_long() && mask_t->lo_as_long() >= 0) {
+    return true;
+  }
+
+  return false;
+}
+
+// Helper method to transform:
+// patterns similar to (v1 + (v2 << 2)) & 3 to v1 & 3
+Node* MulNode::AndIL_add_shift_and_mask(PhaseGVN* phase, BasicType bt) {
```

Comment:
Similar comments for this function also.

```
// Given an expression (AndX (AddX v1 (LShiftX v2 #N)) #M)
// determine if the AndX must always produce (AndX v1 #M),
// because the shift (v2<<N) is bitwise disjoint from the mask #M.
// The X in AndX will be I or L, depending on bt.
// Specifically, the following cases fold,
// when the shift value N is large enough to zero out
// all the set positions of the and-mask M.
//   (AndI (AddI v1 (LShiftI _ #N)) #M) => v1
//   (AndL (AddI v1 (LShiftL _ #N)) #M) => v1
//   (AndL (AddL v1 (ConvI2L (LShiftI _ #N))) #M) => v1
// The M and N values must satisfy ((-1 << N) & M) == 0.
// Because the optimization might work for a non-constant
// mask M, and because the AddX operands can come in either
// order, we check for every operand order.
Node* MulNode::AndIL_add_shift_and_mask(PhaseGVN* phase, BasicType bt) {
  Node* add  = in(1);
  Node* mask = in(2);
+ if (add == NULL || mask == NULL) {
+   return NULL;
+ }
+ int addidx = 0;
  if (add->Opcode() == Op_Add(bt)) {
+   addidx = 1;
+ } else if (mask->Opcode() == Op_Add(bt)) {
+   mask = add;
+   addidx = 2;
+   add = in(addidx);
+ }
+ if (addidx > 0) {
    Node* add1 = add->in(1);
    Node* add2 = add->in(2);
    if (add1 != NULL && add2 != NULL) {
      if (AndIL_shift_and_mask(phase, add1, mask, bt, false)) {
        set_req_X(addidx, add2, phase);
        return this;
      } else if (AndIL_shift_and_mask(phase, add2, mask, bt, false)) {
        set_req_X(addidx, add1, phase);
        return this;
      }
    }
  }
  return NULL;
}
```